README.md
cataloniaframework_stable_v1
============================

Catalonia Framework - Stable release - V1.1.017 released on 2014-09-02


doc/changelog.txt
2015-08-28 - v.1.1.017
Added composer.json

Added phpunit to composer

Added an empty vendor/autoload.php

Removed a require_once to init/bootstrap.php

2015-08-27 - v.1.1.017
Added more debug info when in MySqli trying to change charset and failing.

Added a trap error when connect to Mysqli failed. Before was not detected as it, but to a charset problem.

2015-08-23 - v.1.1.017
Just quickly finished CUD implementation. Job pending.

------------------------------------------------------------------

2014-04-21 - v.1.1.016
Carles Mateo
Added improved config for C.U.D.:
    // Use for Cassandra Universal Driver
    'cassandra_cud_hostname'            => '127.0.0.1',
    'cassandra_cud_port'                => '80',
    'cassandra_cud_cluster_hostname'    => '127.0.0.1',
    'cassandra_cud_cluster_port'        => '9906',
    'cassandra_cud_cluster_username'    => '',
    'cassandra_cud_cluster_password'    => ''


2014-04-10 - v.1.1.016
Carles Mateo
Fixed bug in enconding KeySpace to C.U.D.

2014-04-09 - v.1.1.016
Carles Mateo
Added support for Cassandra Universal Driver

Added new vars to config:
    // Use for Cassandra Universal Driver
    'cassandra_cud_hostname' => '127.0.0.1',
    'cassandra_cud_port'     => '9906',
    'cassandra_cud_username' => '',
    'cassandra_cud_password' => ''

2014-02-01 - v.1.1.016
Carles Mateo
Sections now support mapping urls based on two blocks, like:
/en/personal-folder/my-files

2014-01-31 - v.1.1.016
Carles Mateo
init/bootstrap.php is split in two files:
init/customprebootstrap.php that is loaded at a very early stage, before any custom method
init/custompostbootstrap.php that is loaded at the end of bootstrap.php, when all the variables have been set.

2014-02-01 - v.1.1.016
Carles Mateo
Added class to core Graphics

Added method to draw a flexible stats graphic

Improved Datetime::getDateTime now accepts a different time.
So you can use the method to calculate the datetimes from other days, e.g. yesterday: time()-(24*60*60)

------------------------------------------------------------------

2014-01-29 - v.1.1.015 Released
Created crons/ folder

2014-01-28 - v. 1.1.015
File::deleteFile now returns true on success false of failure

getter Db->getCqlPath

getter and setter for UseDatabaseOrKeyspace
New functionality to do not perform use database or keyspace, specially useful for Database Admin operations. (ignored in POSTGRE)

2014-01-27 - v. 1.1.015
Fixed issue, query_type was not set in the Array of results from Db.

All the CQLSÍ generated files in /tmp begin with cqlsi- in order to make it easier to list, or delete.

Added public var to Db Class to point where cqlsh is.

Added Db->setCqlPath for easier Path changing if cqlsh is not under the default directory

Db for Cassandra, first row is Array index 0, not 1, to act the same as other drivers in Db

2014-01-26 - v. 1.1.015
Added REGEXP_EMAIL to Form class

Added CommonRequests::initSession
This function will be called before any other custom in commonrequests.
This helps to create Login and Session functionality

Fixed issue with field names in CQLSÍ, that were added with ending spaces. Have been trimmed.

------------------------------------------------------------------

2014-01-11 - v. 1.1.014 Released
Bug fixed: port was incorrect for server_port in config for production and preproduction.
If the port was the default nothing was affected.

2014-01-10 - v. 1.1.014
Finished CQLSÍ Cassandra driver interface

Cassandra can keep all the Database commands and results.

2014-01-09 - v. 1.1.014
Added query_for_driver to Database results.

Db::prepareInsert contemplates CASSANDRA

2014-01-08 - v. 1.1.014
Carles Mateo

Created Extended UUID Security::EUUID

Added NODE configuration for Load Balancers

Added Class File with:
    - logToFile, able to write strings or objects and other types
    - writeToFile

------------------------------------------------------------------

2014-01-08 - v. 1.1.013 Released
Carles Mateo:
Headers are sent always in index.php

Improved first-time template

Added a mechanism to track new installations of the Framework

Improved documentation

2014-01-07 - v. 1.1.013
Carles Mateo:
Created CommonRequests::logRequest

Created CommonRequests::endRequest

Added Contact Form for DemoApp in English Template

Added RESPONSE_IMAGE_PNG

Fixed prefix st_ that was appearing as s_

2014-01-06 - v. 1.1.013
Added constant for LIB folder

Security class loaded in bootstrap

Improvements in Security Class to BASE64 and added Control String for ciphered fields.

Fixed a bug when raising a CustomException from autoload (was out of the namespace).

cipherParam now allow to crypt and decrypt objects (as well as binary).

2014-01-05 - v. 1.1.013
Added Security Class with cipher and decipher for params, and UUID v3, v4 and v5.

------------------------------------------------------------------

2014-01-01 - v. 1.1.012 Released
Added jquery-1.10.2 to DemoApp

2013-12-31 - v. 1.1.012
Changed default locale to en_US.UTF8

2013-12-29 - v. 1.1.012
Added Custom Exception to index.php to differentiate between database error (if developer wants)

2013-12-28 - v. 1.1.012
Improved Requests::getClientIp

Added new Custom Exception:
class DatabaseUnableToSetCharset extends Exception {}

Fixed bug with port for MySqli driver in Db class

2013-12-25 - v. 1.1.012
Autoload improved

Bootstrap improved to allow multi-development environments (requests.class.php is loaded first)

------------------------------------------------------------------

2013-12-23 - v. 1.1.011 - Released version
Fixed feature that only worked with PHP 5.4 and greater to PHP 5.3

New method in Requests class getHttpReferer

New method Requests::getServerName

2013-12-22 - v. 1.1.011
Added support for right bar at getSectionUrl

Added default empty params to sample actions to make it clear and solve frequent asked questions

Released version, download from:
https://github.com/cataloniaframework/cataloniaframework_stable_v1


catfwcore/bootstrap.php
<?php
/**
 * Creator:      Carles Mateo
 * Date:         2013-02-07 12:53
 * Last Updater: Carles Mateo
 * Last Updated: 2015-09-03 13:10
 * Filename:     bootstrap.php
 * Description:  The bootstrap file
 */

// This is to load the autoload from the composer.json
require_once '../vendor/autoload.php';

require_once 'requests.class.php';  // For Multi-Development environments in development config
use CataloniaFramework\Requests as Requests;

require_once '../config/general.php';
// Db is required in general
use CataloniaFramework\Db as Db;

// Define Autoload
require_once 'autoload.php';

// Register our CatFw Autoload
spl_autoload_register('catfw_autoload');

require_once CATFW_CORE_ROOT.'strings.class.php';
use CataloniaFramework\Strings as Strings;
require_once CATFW_CORE_ROOT.'section.class.php';
use CataloniaFramework\Section as Section;
require_once CATFW_CORE_ROOT.'core.class.php';
use CataloniaFramework\Core as Core;
require_once CATFW_CORE_ROOT.'core.php';
require_once CATFW_CORE_ROOT.'controllerbase.class.php';
use CataloniaFramework\ControllerBase as ControllerBase;
require_once CATFW_CORE_ROOT.'modelbase.class.php';
require_once CATFW_CORE_ROOT.'views.class.php';
use CataloniaFramework\Views as Views;
require_once CATFW_CORE_ROOT.'datetime.class.php';
use CataloniaFramework\DateTime as DateTime;
require_once CATFW_CORE_ROOT.'customexceptions.class.php';
require_once CATFW_CORE_ROOT.'navigation.class.php';
use CataloniaFramework\Navigation as Navigation;
require_once CATFW_CORE_ROOT.'section.class.php';
require_once CATFW_CORE_ROOT.'menu.class.php';
use CataloniaFramework\Menu as Menu;
require_once CATFW_CORE_ROOT.'translations.class.php';
use CataloniaFramework\Translations as Translations;
require_once CATFW_CORE_ROOT.'numbers.class.php';
use CataloniaFramework\Numbers as Numbers;
require_once CATFW_CORE_ROOT.'currency.class.php';
use CataloniaFramework\Currency as Currency;
// Register common to all pages, f.e. user vars
require_once CUSTOM_INIT_ROOT.'commonrequests.class.php';
use CataloniaFramework\CommonRequests as CommonRequests;
require_once CATFW_CORE_ROOT.'form.class.php';
use CataloniaFramework\Form as Form;
require_once CATFW_CORE_ROOT.'session.class.php';
use CataloniaFramework\Session as Session;
require_once CATFW_CORE_ROOT.'security.class.php';
use CataloniaFramework\Security as Security;
require_once CATFW_CORE_ROOT.'file.class.php';
use CataloniaFramework\File as File;
require_once CATFW_CORE_ROOT.'graphics.class.php';
use CataloniaFramework\Graphics as Graphics;

$s_user_language = LANGUAGE_DEFAULT;

$o_db = new Db($st_server_config['database']);

$b_is_custom_section = false;

$s_params = Requests::getParamStringGET('params');
$st_params = Strings::getParamsFromURL($s_params);

define('REQUESTED_PATH', $s_params);

require_once CUSTOM_INIT_ROOT.'customprebootstrap.php';

CommonRequests::initSession($o_db);

// Define user custom hardcoded URLs
CommonRequests::registerURLS();

if (Navigation::isURLCustom(REQUESTED_PATH) == true) {
    // We use a special method to load Custom Content
    $s_controller = 'Index';
    $s_action = 'CATFW_getCustomContent';

    if (!Core::loadController($s_controller)) {
        echo getErrorView(Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND, Views::$st_ERROR_MESSAGES[Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND]);
        Core::end();
    }

    Core::loadController($s_controller);

    $s_controller_with_namespace = '\\CataloniaFramework\\'.$s_controller;
    $o_controller = new $s_controller_with_namespace;

} else {
    // Is not custom url, so we will get the class and method to execute

    // This is a special index for the case of the basic domain call without any controller specified
    $s_controller = 'Index';
    // Default Action name
    $s_action = 'Index';

    if (MULTILANG == true) {
        if (isset($st_params[0])) {
            if (Core::isLanguageActive($st_params[0]) == false) {
                Core::redirectUserToUrl('/'.$s_user_language.'/');
                Core::end();
            }
            // The URL language is correct
            $s_user_language = $st_params[0];
            define('USER_LANGUAGE', $s_user_language);
            unset($st_params[0]);
        } else {
            // Got http://domain/ but the right thing would be http://domain/lang/ , so redirect
            Core::redirectUserToUrl('/'.$s_user_language.'/');
            Core::end();
        }

        // Register user sections
        CommonRequests::registerSections($o_db);

        if (isset($st_params[1]) && !empty($st_params[1])) {

            $st_possible_section = null;
            if (isset($st_params[2]) && !empty($st_params[2])) {
                // Check first for Section of two levels. E.g. /en/dashboard/main
                $st_possible_section = Section::getSectionInfoByPath('/'.$s_user_language.'/'.$st_params[1].'/'.$st_params[2]);
                if ($st_possible_section !== null) {
                    // Section Found
                    $b_is_custom_section = true;
                    $s_controller   = $st_possible_section['controller'];
                    $s_action       = $st_possible_section['action'];
                }
            }

            if ($st_possible_section == null) {
                // Check for matching Section of one level. E.g. /en/dashboard
                $st_possible_section = Section::getSectionInfoByPath('/'.$s_user_language.'/'.$st_params[1]);

                if ($st_possible_section !== null) {
                    // Section Found
                    $b_is_custom_section = true;
                    $s_controller   = $st_possible_section['controller'];
                    $s_action       = $st_possible_section['action'];
                } else {
                    // Common structure /lang/controller/action
                    if (isset($st_params[1]) && !empty($st_params[1])) {
                        $s_controller = Strings::getSanitizedControllerName($st_params[1]);
                        unset($st_params[1]);
                    }
                    if (isset($st_params[2]) && !empty($st_params[2])) {
                        $s_action = Strings::getSanitizedControllerName($st_params[2]);
                        unset($st_params[2]);
                    }
                }

            }

        }

        Section::registerSection('index', '/'.$s_user_language.'/');
    } else {

        define('USER_LANGUAGE', LANGUAGE_DEFAULT);
        if (isset($st_params[0]) && !empty($st_params[0])) {
            $st_possible_section = Section::getSectionInfoByPath($st_params[0]);
            if ($st_possible_section !== null) {
                // Section Found
                $b_is_custom_section = true;
                $s_controller   = $st_possible_section['controller'];
                $s_action       = $st_possible_section['action'];
            } else {
                // Common structure /controller/action
                if (isset($st_params[0]) && !empty($st_params[0])) {
                    $s_controller = Strings::getSanitizedControllerName($st_params[0]);
                    unset($st_params[0]);
                }
                if (isset($st_params[1]) && !empty($st_params[1])) {
                    $s_action = Strings::getSanitizedControllerName($st_params[1]);
                    unset($st_params[1]);
                }
            }
        }

        Section::registerSection('index', '/', 'Index', 'Index');
    }

    // Check security for controller and name
    if (!Core::isValidName($s_controller) || !Core::isValidName($s_action) ||
        !Core::loadController($s_controller)) {
        // Invalid name of controller or action
        echo getErrorView(Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND, Views::$st_ERROR_MESSAGES[Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND]);
        Core::end();
    }

    // Load the Controller Class into memory
    $s_controller_with_namespace = '\\CataloniaFramework\\'.$s_controller;
    $o_controller = new $s_controller_with_namespace;

    // For example: actionIndex
    $s_action = 'action'.$s_action;

    if (!method_exists($o_controller, $s_action)) {
        // Invalid name of controller or action
        echo getErrorView(Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND, Views::$st_ERROR_MESSAGES[Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND]);
        Core::end();
    }

}

define('CONTROLLER', $s_controller);
define('ACTION', $s_action);

// Create parameter list
$st_params = array_values($st_params);
$st_params_url = Array();
for ($i_count=0; $i_count<=count($st_params); $i_count = $i_count+2) {
    if (isset($st_params[$i_count+1])) {
        $st_params_url[$st_params[$i_count]] = $st_params[$i_count +1];
    }
}

// Load Additional custom content
CommonRequests::registerUserVars($o_db);

// Log the visit
CommonRequests::logRequest($o_db);

require_once CUSTOM_INIT_ROOT.'custompostbootstrap.php';
